lib/rucio/client/uploadclient.py

```diff
@@ -129,15 +130,22 @@ def upload(self, items, summary_file_path=None, traces_copy_out=None):
         # and cache rse settings
         registered_dataset_dids = set()
         registered_file_dids = set()
+        rses = []
+        rse_expression = None
         for file in files:
-            rse = file['rse']
+            rse_expression = file['rse']
+            if not rses:
+                rses = [r['rse'] for r in self.client.list_rses(rse_expression)]  # can raise InvalidRSEExpression
+                random.shuffle(rses)
```

Comment:
this is weird, but i guess that's what it's supposed to do. @bari12 opinion?

Reply:
I don't think this entirely works. The `files` list might have files where the rse is already set. In this case the algorithm overwrites it. I think it needs to be a bit smarter than that. 
But it also shouldn't re-evaluate the rse_expression (if it is the same) over and over again. So a bit more logic is needed here.